### PR TITLE
Shipping Labels: customs form UI part 2

### PR DESCRIPTION
**Pull request by hichamboushaba**
Third PR of #3526, this PR adds UI interactions and validation. The changes aim to have a unidirectional flow of data from the ViewModel to the UI, so each change to the UI is passed to the ViewModel, then it's reflected in the items' list using the DiffUtil.
Please note that the customs form is still disconnected from the rest of the steps, and uses only fake data (except the destination country which is passed to allow validation testing)

Please make sure #4077 is merged before merging this.

#### Tests
For all the following tests, please use an order that's eligible for shipping label creation, and pass the initial steps until your reach the customs step.
##### ITN validation 1
1. Make sure that a country not included in the [`USPS_ITN_REQUIRED_DESTINATIONS`](https://github.com/woocommerce/woocommerce-android/blob/d316b0c17dd5781e1cf56ed5c8970323e28dc927/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt#L32) is selected in the destination address step.
2. Type anything in the ITN field.
3. Confirm that "invalid format" is displayed.
4. Type "AES X20160406131357" in the field.
5. Confirm that the error is hidden.

##### ITN validation 2
1. Make sure that a country not included in the [`USPS_ITN_REQUIRED_DESTINATIONS`](https://github.com/woocommerce/woocommerce-android/blob/d316b0c17dd5781e1cf56ed5c8970323e28dc927/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt#L32) is selected in the destination address step.
2. Clear the ITN field.
3. Type "123456" in the hs tariff field of one of the items.
4. Change the value field to "2501"
5. Confirm that the ITN field shows the error: "International Transaction Number is required for shipping items valued over $2,500..."
6. Type "AES X20160406131357" in the field.
7. Confirm that the error is hidden.

##### ITN validation 3
1. Make sure that one of the countries of this list [`USPS_ITN_REQUIRED_DESTINATIONS`](https://github.com/woocommerce/woocommerce-android/blob/d316b0c17dd5781e1cf56ed5c8970323e28dc927/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt#L32) is selected in the destination address step.
2. Clear the ITN field.
3. Confirm that the ITN field shows the error: "International Transaction Number is required for shipments..."
6. Type "AES X20160406131357" in the field.
7. Confirm that the error is hidden.

##### HS tariff validation
1. Confirm that an empty hs tariff number doesn't show any error.
2. Type a number less than 6 digits
3. Confirm that the field shows the error: "The tariff number must be 6 digits long"
4. Type 6 digits.
5. Confirm that the error is removed.

##### Contents type validation
1. Select "Other" for the contents type.
2. Confirm that the description field is displayed.
2. Confirm that the error: "Please describe what kind of goods this package contains." is displayed in the field.
3. Type something.
4. Confirm that the error is removed.

##### Restriction type validation
1. Select "Other" for the restriction type.
2. Confirm that the description field is displayed.
2. Confirm that the error: "Please describe what kind of restrictions this package must have." is displayed in the field.
3. Type something.
4. Confirm that the error is removed.

##### Item description validation
1. Clear one of the items' description.
2. Confirm that the error: "This field is required" is displayed.

##### Weight validation
1. Clear one of the items' weight
2. Confirm that the error: "This field is required" is displayed.
3. Type 0 in the field.
4. Confirm that the error: "Weight must be greater than zero" is displayed.
5. Type something greater than 0.
6. Confirm that the error is removed.

##### Value validation
1. Clear one of the items' value
2. Confirm that the error: "This field is required" is displayed.
3. Type 0 in the field.
4. Confirm that the error: "Declared value must be greater than zero" is displayed.
5. Type something greater than 0.
6. Confirm that the error is removed.

##### Done button visibility
Make sure that when any validation rule above is broken, the "done" button gets hidden.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsAdapter.kt [hunk 1]
```diff
@@ -61,30 +96,109 @@ class ShippingCustomsAdapter : RecyclerView.Adapter<PackageCustomsViewHolder>()
...
+            val diffResult = DiffUtil.calculateDiff(CustomsLineDiffCallback(field, value))
             field = value
-            notifyDataSetChanged()
+            diffResult.dispatchUpdatesTo(this)
         }
 
+    var parentItemPosition: Int = -1
+
+    override fun getItemId(position: Int): Long {
+        return position.toLong()
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsAdapter.kt]: Above we have `setHasStableIds(true)`, but can we be sure that `position` is going to be unique if the list of items changes? It looks like we can use `itemId` of the `CustomsLine` as the unique ID.

**Comment by nbradbury** [comment]
Was this error text in the designs? It seems very wordy. If nothing else, we could shorten "International Transaction Number" to "ITN" since the "Learn more" label includes "International Transaction Number." My vote, though, would be to not include the message about the specific tariff number, so: "ITN is required for shipping items valued over $2,500 per tariff number."

Screenshot_1622207134

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsAdapter.kt (hunk 1 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsAdapter.kt]: The position won't be changed for this screen, as the user doesn't have a way to change the packages or their contents at this step, it's all done in the packages step before.
But I agree with you that using the `itemId` is better, I'll update it.

**Comment by hichamboushaba** [comment]
> Was this error text in the designs? It seems very wordy. If nothing else, we could shorten "International Transaction Number" to "ITN" since the "Learn more" label includes "International Transaction Number." My vote, though, would be to not include the message about the specific tariff number, so: "ITN is required for shipping items valued over $2,500 per tariff number."
> 
> Screenshot_1622207134

I took all the error messages from the web client, as those cases are not covered in the Figma file, and I was planning to validate all those cases in the final PR with the design team.
I'll shorten the message for this case, and get design's team feedback once the feature is actually completed 🙂

File: WooCommerce/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -507,9 +507,13 @@
     <string name="shipping_label_customs_contents_type_other_hint">Contents details</string>
     <string name="shipping_label_customs_restriction_type_other_hint">Restriction details</string>
     <string name="shipping_label_customs_itn_hint" translatable="false">ITN</string>
+    <string name="shipping_label_customs_itn_invalid_format">Invalid format</string>
+    <string name="shipping_label_customs_itn_required_items_over_2500">International Transaction Number is required for shipping items valued over $2,500 per tariff number. Products with tariff number %1$s add up to more than $2,500.</string>
+    <string name="shipping_label_customs_itn_required_country">International Transaction Number is required for shipments to %1$s.</string>
```
Comment (nbradbury) [inline WooCommerce/src/main/res/values/strings.xml]: np: Google's [guidelines](https://material.io/design/communication/writing.html#principles) recommend not using periods for single sentences. Also, I think we can shorten this to "ITN."

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt [hunk 3]
```diff
@@ -2,74 +2,337 @@ package com.woocommerce.android.ui.orders.shippinglabels.creation
...
 import com.woocommerce.android.viewmodel.ScopedViewModel
+import com.woocommerce.android.viewmodel.navArgs
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.parcelize.IgnoredOnParcel
 import kotlinx.parcelize.Parcelize
+import org.wordpress.android.fluxc.store.WCDataStore
 import java.math.BigDecimal
 import javax.inject.Inject
 
+private val ITN_REGEX = Regex("""^(?:(?:AES X\d{14})|(?:NOEEI 30\.\d{1,2}(?:\([a-z]\)(?:\(\d\))?)?))${'$'}""")
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt:34]: WDYT about adding a comment to `ITN_REGEX` and `HS_TARIFF_NUMBER_REGEX` which describes what the format should be?

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt:34]: I'll be happy to do it, but I don't it can be easily summarized in a comment, especially for the ITN (for the HS number it's pretty clear it's just a 6 digits number).
The ITN has so many variations (check Appendix A of this file: https://mailomg.files.wordpress.com/2018/05/exportcompliancecustomsdatarequirements.pdf), so what I did is that I copied the regex from the plugin's source code.
WDYT?

**Comment by nbradbury** [comment]
I'm seeing the ITN warning even when I enter a value less than $2500. Am I misunderstanding the warning?

Screenshot_1622209963

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt (hunk 3 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt:34]: Got it! Maybe just add a comment with that link in it for future developers?

**Comment by hichamboushaba** [comment]
> I'm seeing the ITN warning even when I enter a value less than $2500. Am I misunderstanding the warning?
> 
> Screenshot_1622209963

The ITN warning groups items that have the same `hs tariff` number, and then calculates the sum of `value * quantity` of each item, so for your case, the cause is that the first item in the package has a `quantity = 5`, so the sum is higher than $2500.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingCustomsViewModel.kt:34]: Done in e5e1f92

**Review by nbradbury — APPROVED** [review]
:shipit:
